### PR TITLE
ROB : cope with invalid length in streams

### PR DESCRIPTION
**Pull request by pubpub-zz**
issue https://github.com/py-pdf/PyPDF2/issues/301

in case of invalid extract stream data looking for endstream

### CODE REVIEW

File: PyPDF2/generic.py [hunk 1]
```diff
@@ -578,6 +578,29 @@ def writeToStream(self, stream, encryption_key):
...
+                if p1 > l[o] and p < l[o]:
+                    p1 = l[o]
+            if len(remGens) == 1:
+                return p1
+            else:
+                return getNextObjPos(p, p1, remGens[1:], pdf)
+
+        def readUnsizedFromSteam(stream, pdf):
+            # we are just pointing at beginning of the stream
+            eon = getNextObjPos(stream.tell(), 2**32, [g for g in pdf.xref], pdf) - 1
```
Comment (MartinThoma) [inline PyPDF2/generic.py:593]: what does "eon" mean?

File: PyPDF2/generic.py [hunk 2]
```diff
@@ -657,6 +681,10 @@ def readFromStream(stream, pdf):
                 if end == b_("endstream"):
                     # we found it by looking back one character further.
                     data["__streamdata__"] = data["__streamdata__"][:-1]
+                elif not pdf.strict:
+                    stream.seek(pstart, 0)
+                    data["__streamdata__"] = readUnsizedFromSteam(stream, pdf)
+                    pos = stream.tell()
```
Comment (MartinThoma) [inline PyPDF2/generic.py:687]: Let me check my understanding: If this part is reached, then we are in a situation in which the "endstream" marker was not found where it should be. As we are in best effort mode, we go back to the very beginning of the stream. From there, we read as much as necessary. Right?

File: PyPDF2/generic.py [hunk 3]
```diff
@@ -578,6 +578,29 @@ def writeToStream(self, stream, encryption_key):
...
+            l = pdf.xref[remGens[0]]
+            for o in l:
+                if p1 > l[o] and p < l[o]:
+                    p1 = l[o]
+            if len(remGens) == 1:
+                return p1
+            else:
+                return getNextObjPos(p, p1, remGens[1:], pdf)
+
+        def readUnsizedFromSteam(stream, pdf):
```
Comment (MartinThoma) [inline PyPDF2/generic.py:591]: I don't understand this function. How does it know where the ream ends? Couldn't there be multiple objects in a stream?

Can you point me to a resource where I can read up on PDF streams in general? I would appreciate that a lot :pray:

File: PyPDF2/generic.py (hunk 1 above)
Comment (pubpub-zz) [reply PyPDF2/generic.py:593]: end before next object : we have to find when to stop looking for endstream

File: PyPDF2/generic.py (hunk 2 above)
Comment (pubpub-zz) [reply PyPDF2/generic.py:687]: we read up to the endstream tag

**Comment by MartinThoma** [comment]
You're on fire! Pretty nice! I'll try to understand + mehr it to the 1.x version tomorrow

File: PyPDF2/generic.py (hunk 3 above)
Comment (pubpub-zz) [reply PyPDF2/generic.py:591]: From my analysis/tests, the stream were most of the time compressed/encoded 
Also we have to remember that this function will  only be used to load some data when the PDF file is corrupted.

I peek up some information in the standard
https://opensource.adobe.com/dc-acrobat-sdk-docs/standards/pdfstandards/pdf/PDF32000_2008.pdf
and also in those:
https://www.adobe.com/technology/pdfs/presentations/KingPDFTutorial.pdf
https://www.oreilly.com/library/view/developing-with-pdf/9781449327903/ch01.html

for ref also so test files
https://github.com/pdf-association/pdf20examples

**Pull request merged**